packages/patterns/docs/marshal-vs-patterns-level.md

```diff
@@ -0,0 +1,94 @@
+|            | Float64       | `number`           | `number`      | Only one `0.0`, Only one `NaN` |
+|            | SignedInteger | `bigint`           | `bigint`      |                  |
+|            | Symbol        | `symbol`           | `symbol`      | well-known & registered only (names TBD) |
+|            | String        | `string`           | `string`      | surrogate confusion (TBD) |
+|            | ByteString    | `byteString` (TBD) | `object`      | UInt8Array (TBD) |
+| Containers |               |                    |               |                  |
+|            | Sequence      | `copyArray`        | `object`      | Array            |
+|            | Struct        | `copyRecord`       | `object`      | POJO             |
+|            | Tagged        | `tagged`           | `object`      | Tagged           |
+| Capability |               |                    |               |                  |
+|            |               | `remotable`        | `function`    | Far function     |
+|            |               | `remotable`        | `object`      | Far object with methods |
+|            |               | `remotable`        | `object`      | Remote presence  |
+|            |               | `promise`          | `object`      | Promise          |
+| Others     |               |                    |               |                  |
+|            | Error         | `error`            | `object`      | Error            |
+
+The `@endo/marshal` package defines encodings of the data model for purposes of serialization and transmission over such protocols. The `@endo/marshal` package also defines a rank-ordering over all passable values, where a "rank-order" is a full-order with ties. Tied values are said to be of the *same rank*. A rank-order can be used for sorting. This rank-oder itself is not intended to make sense to the application programmer.
+
+The `@endo/patterns` package defines the `kindOf` taxonomy, which includes additional containers, Keys and Patterns, and a `compareKeys` partial-order over keys that is designed to be meaningful and useful to the applications programmer.
+
+## `kindOf` *vs* `passStyleOf`
+
+For every passable value except Tagged, `kindOf(p) === passStyleOf(p)`. They differ only when `passStyleOf(p) === 'tagged'`. For those, either the `kindOf` level recognizes the Tagged as encoding one of the higher level types defined by the `kindOf` level, or it does not. If it does not, then `kindOf(p) === undefined`. Only the `passStyleOf` level is assumed for universal interoperability. Different participants may encode and recognize different extensions using the same Tagged extension point. If a participant that does not recognize a particular Tagged according to a higher layer it uses, it must still treat it as a valid Tagged object, which must still round trip. Alice might send a Tagged she recognizes to Bob, who does not, but who sends it on to Carol who does.
```

Comment:
I don't understand this. We do apply full rank orderings to unrecognized tag types, in that `rankCompare` just does lexicographic ordering of Taggeds, independent of any higher level recognition. Do I misunderstand the question?

Reply:
Re https://github.com/endojs/endo/issues/1739 , from that thread, I think we resolved that we should just use unicode / utf-8 ordering for rankCompare, and therefore keyCompare, and avoid using UTF-16 / CESU-8 ordering for anything at these levels of abstraction. Does that help clarify anything here?

To make that switch, we'll need to ensure that our stores (heap, virtual, and durable) sort their keys in a way that remains consistent with rankCompare. This may conflict with existing durable data, which will be an interesting problem.

Attn @FUDCo @warner